### PR TITLE
feat(verification-code): add a expiration and resending to verification codes

### LINKED ISSUE #593: [BACKEND] Implement email verification code max lifetime
**Is your feature request related to a problem? Please describe.**

We should add a maximum lifetime of the email verification code to prevent bruteforce. Should also look into possibly adding rate limits to the registration route to prevent API abuse. 

You will likely edit the `User` schema to hold this new data, and modify the email verification code to have a cooldown for resending verification codes, and also the register route to ensure that the code is still valid.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**BEFORE MERGING**
- [ ] Integration test written for services
- [ ] [Schemas annotated](https://payloadcms.com/docs/configuration/collections) if adding new models
- [ ] Code generation run (*hint*: `pnpm generate:types`)
- [ ] Appropriate mocks created where possible
- [ ] PR Reviewed (For non-trivial changes)
- [ ] Changes tested after rebasing on master or merging in master (*hint*: `git fetch origin master:master`, then `git rebase master` or `git merge master`)
- [ ] All required PR checks passing

### PR DESCRIPTION
**Pull request by mikais13**
# Description

Updates `User` config, and relevant routes to use an expiry time of 10 minutes, and a cooldown period between codes of 5 minutes.

Fixes #593 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [ ] Manual testing (requires screenshots or videos)
- [x] Integration tests written (requires checks to pass)

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added thorough tests that prove my fix is effective and that my feature works
- [x] I've requested a review from another user

### CODE REVIEW

File: apps/backend/src/app/api/auth/register/route.test.ts [hunk 1]
```diff
@@ -15,11 +15,21 @@ describe("tests /api/auth/register", () => {
     emailVerificationCode: "123456",
   } satisfies RegisterRequestBody
 
+  const now = new Date()
+  const createdAt = now.toISOString()
+  const expiresAt = new Date(now.getTime() + 24 * 60 * 60 * 1000).toISOString()
+
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/auth/register/route.test.ts:21]: there needs to be an exported `const` for the expiry time

Comment (jeffplays2005) [reply apps/backend/src/app/api/auth/register/route.test.ts:21]: i also think 24 hours for a verification code is too long, usually a few minutes is good. I'd recommend 10 minutes!

File: apps/backend/src/app/api/auth/register/route.ts [hunk 2]
```diff
@@ -27,9 +27,23 @@ export const POST = async (req: NextRequest) => {
...
+    let user = await userDataService.getUserByEmail(email)
 
-    if (user?.emailVerificationCode !== emailVerificationCode) {
+    let verified = false
+    const now = new Date()
+    for (const verification of user.emailVerification || []) {
+      if (
+        verification.verificationCode === emailVerificationCode &&
+        new Date(verification.expiresAt) > now
+      ) {
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/auth/register/route.ts:38]: tbh I think `.some` is easier to read in this case, you can use something like this:

```ts
user.emailVerification?.some(code => verification.verificationCode === emailVerificationCode &&
        new Date(verification.expiresAt) > now)
```

File: apps/backend/src/app/api/auth/register/route.ts [hunk 3]
```diff
@@ -27,9 +27,23 @@ export const POST = async (req: NextRequest) => {
...
+        user = await userDataService.updateUser(user.id, {
+          emailVerification: [],
+        })
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/auth/register/route.ts:41]: just realised there is no need for register route to give `user` as part of it's response body, i've made a ticket for it to be removed! https://github.com/UoaWDCC/uabc-web/issues/745

File: apps/backend/src/app/api/auth/verification-code/route.test.ts [hunk 4]
```diff
@@ -53,7 +70,59 @@ describe("/api/auth/verification-code", () => {
...
+
+    // expect(res.status).toBe(StatusCodes.OK)
+    // expect(await res.json()).toEqual({ message: "Verification code sent" })
+    // const user = await userDataService.getUserByEmail(email)
+    // expect(user.emailVerification).toContainEqual({
+    //   id: expect.any(String),
+    //   verificationCode: code,
+    //   createdAt: expect.any(String),
+    //   expiresAt: expect.any(String),
+    // })
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/auth/verification-code/route.test.ts:102]: remove commented out code!

**Review by jeffplays2005 — COMMENTED** [review]
Overall a good implementation although a couple questions required to address:

- It's good to implemented a list of verification codes, although for good convention, we only want the latest code to work (the rest are irrelevant). 
- What happens after 10 verification codes? (low likelihood of occurring but could happen)
  - Could potentially revamp the logic on how the current system work (could use the list as a queue and see the latest last requested code, if too early, then can deny the request, no longer needs a system of 10), systems tend to use a 15 minute cool down or something on re-requesting codes!

**Comment by mikais13** [comment]
> Overall a good implementation although a couple questions required to address:
> 
> * It's good to implemented a list of verification codes, although for good convention, we only want the latest code to work (the rest are irrelevant).
> * What happens after 10 verification codes? (low likelihood of occurring but could happen)
>   
>   * Could potentially revamp the logic on how the current system work (could use the list as a queue and see the latest last requested code, if too early, then can deny the request, no longer needs a system of 10), systems tend to use a 15 minute cool down or something on re-requesting codes!

I understand what you are saying, but I feel that worsens UX.

Usually, when I resend a verification code, it's because the email is taking a while to send and I want to make sure that I get the email. When we only accept the latest verification code, it means that if a user only sees the first email (i.e. the second code hasn't come through yet), they will attempt to use the code, get denied, probably double check that they didn't mistransfer the code, and try again, before ever rechecking their email (especially since you don't get notifications when your email app is already open).

**Comment by jeffplays2005** [comment]
> > Overall a good implementation although a couple questions required to address:
> > 
> > * It's good to implemented a list of verification codes, although for good convention, we only want the latest code to work (the rest are irrelevant).
> > * What happens after 10 verification codes? (low likelihood of occurring but could happen)
> >   
> >   * Could potentially revamp the logic on how the current system work (could use the list as a queue and see the latest last requested code, if too early, then can deny the request, no longer needs a system of 10), systems tend to use a 15 minute cool down or something on re-requesting codes!
> 
> I understand what you are saying, but I feel that worsens UX.
> 
> Usually, when I resend a verification code, it's because the email is taking a while to send and I want to make sure that I get the email. When we only accept the latest verification code, it means that if a user only sees the first email (i.e. the second code hasn't come through yet), they will attempt to use the code, get denied, probably double check that they didn't mistransfer the code, and try again, before ever rechecking their email (especially since you don't get notifications when your email app is already open).

I understand the importance of UX but at the cost of security, I think it’s too much of a risk. Best to leave it at one verification code at a time, I’d rather ensure a reliable mail service than having several active codes at the same time(greater chances of a brute force attack being successful). 

I also think it is intuitive enough that resending a verification code invalidates the previous(quite similar to all the other platforms that use this form of mechanism), the only one being different is MFA which can generate codes at a certain timestamp (hence preventing a user getting confused from entering an expired code (just a second after it re-rolls), by also checking recently expired codes).

File: apps/backend/src/data-layer/utils/DateUtils.test.ts
Comment (jeffplays2005) [inline apps/backend/src/data-layer/utils/DateUtils.test.ts:1]: I think you're missing tests for `getVerificationCodeCoolDownDate`

File: apps/backend/src/data-layer/utils/DateUtils.test.ts [hunk 5]
```diff
@@ -79,4 +83,18 @@ describe("DateUtils", () => {
...
+      const now = new Date()
+      const expiryDate = getVerificationCodeExpiryDate(now)
+
+      expect(expiryDate.getTime()).toEqual(now.getTime() + 10 * 60 * 1000)
+    })
+
+    it("returns a valid Date object", () => {
+      const expiryDate = getVerificationCodeExpiryDate()
+      expect(expiryDate instanceof Date).toBe(true)
+    })
```
Comment (jeffplays2005) [inline apps/backend/src/data-layer/utils/DateUtils.test.ts:98]: i think the test description is a bit off as it's more like the edge case of `getVerificationCodeExpiryDate` using the default `new Date()` when no parameter is passed down to this function

`expect(expiryDate instanceof Date).toBe(true)` should be tested in `"returns a date 10 minutes from now"`

File: apps/backend/src/business-layer/services/MailService.test.ts [hunk 6]
```diff
@@ -18,7 +18,7 @@ describe("MailService", () => {
       expect(sendEmailMock).toHaveBeenCalledWith({
         to: email,
         subject: "Email verification code",
-        text: `Here is your email verification code: ${code}`,
+        text: `Here is your email verification code: ${code}. This code will expire in 24 hours.`,
       })
       expect(result).toEqual({ success: true })
```
Comment (jeffplays2005) [inline apps/backend/src/business-layer/services/MailService.test.ts:23]: should be 10 minutes!

File: apps/backend/src/business-layer/services/MailService.ts [hunk 7]
```diff
@@ -11,7 +11,7 @@ export default class MailService {
     return await payload.sendEmail({
       to: email,
       subject: "Email verification code",
-      text: `Here is your email verification code: ${verificationCode}`,
+      text: `Here is your email verification code: ${verificationCode}. This code will expire in 24 hours.`,
     })
```
Comment (jeffplays2005) [inline apps/backend/src/business-layer/services/MailService.ts:15]: should be 10 minutes!

File: apps/backend/src/app/api/auth/verification-code/route.ts [hunk 8]
```diff
@@ -27,14 +31,42 @@ export const POST = async (req: NextRequest) => {
...
+
+      // check that latest verification code is past cool down period
+      const latestVerification = user.emailVerification?.[0] || null
+      if (
+        latestVerification &&
+        getVerificationCodeCoolDownDate(new Date(latestVerification.createdAt)) > new Date()
+      ) {
+        return NextResponse.json(
+          { error: "A verification code has already been sent recently" },
+          { status: StatusCodes.TOO_MANY_REQUESTS },
```
Comment (jeffplays2005) [inline apps/backend/src/app/api/auth/verification-code/route.ts:43]: would be great to also notify the user that they can request another one such as "Please wait before requesting another one"

File: packages/shared/src/payload-types.ts [hunk 9]
```diff
@@ -374,9 +374,25 @@ export interface User {
...
+         * The current expiration date of this email verification code
+         */
+        expiresAt: string;
+        /**
+         * The date when this email verification code was created
+         */
+        createdAt: string;
+        id?: string | null;
+      }[]
+    | null;
```
Comment (jeffplays2005) [inline packages/shared/src/payload-types.ts:395]: just noticed that it could be `null`, this makes things a bit harder later when testing and actual internal logic, would be best if `required: true` is set in the config! we have a default value of `[]` anyways

**Review by jeffplays2005 — COMMENTED** [review]
Great work!! Just left a couple more comments

**Review by jeffplays2005 — COMMENTED** [review]
Great work but I recommend a single object now that we only have one code at a time!

**Review by jeffplays2005 — APPROVED** [review]
LGTM! Excellent work
